clickstream/src/main/kotlin/clickstream/extension/CSMessageExt.kt

```diff
@@ -73,4 +76,50 @@ public fun MessageLite.eventName(): String? {
         declaredField.isAccessible = true
         declaredField.get(this) as String
     }.getOrNull()
-}
\ No newline at end of file
+}
+
+/**
+ * Converts [MessageLite] to [JSONObject] using reflection.
+ * */
+public fun MessageLite.toJson(): JSONObject {
+
+    fun isValidMethod(method: Method) =
+        method.parameterTypes.isEmpty() && method.name.split(".").last().run {
+            startsWith("get") && !equals("getDefaultInstance") && !contains("Byte")
+        }
+
+    fun getPropertyNameFromMethod(methodName: String) = methodName.split(".").last()
+        .substring(3).substringBefore("(")
+
+    fun getAllPropertyValueAndName(messageLite: MessageLite): List<Property> {
+        val declaredMethods = messageLite.javaClass.declaredMethods
+        val validMethods = declaredMethods.filter { isValidMethod(it) }
+        return validMethods.map {
+            it.isAccessible = true
+            Property(
+                getPropertyNameFromMethod(it.name),
+                it.invoke(messageLite)
+            )
+        }
+    }
+
+    fun getJson(messageLite: MessageLite, json: JSONObject): JSONObject {
+        val listOfProperty = getAllPropertyValueAndName(messageLite)
+        listOfProperty.forEach { property ->
+            val value = when (property.value) {
+                is MessageLite -> getJson(property.value, JSONObject())
+                is List<*> -> property.value.map { propertyValue ->
+                    (propertyValue as? MessageLite)?.run {
+                        getJson(this, JSONObject())
+                    } ?: propertyValue
+                }
+                else -> property.value
+            }
+            json.put(property.name, value)
+        }
+        return json
+    }
+    return getJson(this, JSONObject())
```

Comment:
i am not sure how many allocations we do here. however since the `MessageLite` is an abstract class, can we cast to the origin class?

e.g
```
inline fun <reified T> MessageLite.castTo(): T {
     ...
}
```